### PR TITLE
Foreign function interface (Panama) blog

### PR DESCRIPTION
**Pull request by alanpaxton**
We did some experimental work with FFI and native memory as a potential improvement to the Java API.
The work lives (unmerged) in https://github.com/facebook/rocksdb/pull/11095

This is the report text from that branch, extract as a blog post.
Along with some supporting files (png, pdf of graphs).

### CODE REVIEW

**Comment by alanpaxton** [comment]
> 

Hi @mrambacher For my convenience I just built a wrapper around the C++ which I am familiar with, but you are of course right, FFI means we can easily wrap any C interface into Java, and the way forward for the future should be for RocksDB (however you define it) to maintain that single C interface, and just make sure it supports the features we need from Java, and any other client language that might be needed (Go and Rust spring to mind)

**Comment by alanpaxton** [comment]
@anand1976 rebased and once again ready for review/import.

File: docs/_posts/2023-08-25-foreign-function-interface.markdown [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+
+### Native Memory
+
+Panama's *Foreign-Memory Access API* appears to us to be the most significant part of the whole project. At the `Java` side of RocksDB it gives us a clean mechanism (a `MemorySegment`) for holding RocksDB data (e.g. as from the result of a `get()`) call pending its forwarding to client code or network buffers.
+
+We have taken advantage of this mechanism to provide the core `FFIDB.getPinnableSlice()` method in our Panama-based API. The rest of our prototype `get()` API, duplicating the existing *JNI*-based API, is then a *Pure Java* library on top of `FFIDB.getPinnableSlice()` and `FFIPinnableSlice.reset()`.
+
+The common standard for foreign memory opens up the possibility of efficient interoperation between RocksDB and Java clients (e.g. Kafka). We think that this is really the key to higher performing, more integrated Java-based systems:
+- This could result in data never being copied into Java memory, or a significant reduction in copies, as native `MemorySegment`s are handed off between co-operating Java clients of fundamentally native APIs.
```
Comment (ajkr) [inline docs/_posts/2023-08-25-foreign-function-interface.markdown:223]: is it more error prone? Does the client need something higher level than `MemorySegment` so they can call a cleanup function like `PinnableSlice::reset()`?

**Review by ajkr — APPROVED** [review]
LGTM!

File: docs/_posts/2023-08-25-foreign-function-interface.markdown (hunk 1 above)
Comment (alanpaxton) [reply docs/_posts/2023-08-25-foreign-function-interface.markdown]: The wrapper-ised `get()` variant does this. I have revised the text to be a bit clearer. But it falls out as more of a layered approach, 2-step with `getPinnableSlice()` and `reset()` and the potential for chaining performance, or wrapped with `get()` and other high level methods.

**Pull request closed without merging**